File: queuer/src/main/java/com/rndapp/task_feed/activities/CreateAccountActivity.java
package com.rndapp.task_feed.activities;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.ActionBarActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.google.gson.Gson;
import com.rndapp.task_feed.QueuerApplication;
import com.rndapp.task_feed.R;
import com.rndapp.task_feed.api.ServerCommunicator;
import com.rndapp.task_feed.models.ActivityUtils;
import com.rndapp.task_feed.models.User;
import org.json.JSONObject;

/**
 * Created with IntelliJ IDEA.
 * User: ell
 * Date: 7/13/13
 * Time: 2:28 PM
 */
public class CreateAccountActivity extends ActionBarActivity implements View.OnClickListener{
    private User user;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_account);

        findViewById(R.id.create_account_button).setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        EditText userField = (EditText)findViewById(R.id.username_field);
        EditText passField = (EditText)findViewById(R.id.password_field);
        if (!userField.getText().toString().equals("") && !passField.getText().toString().equals("")){
            findViewById(R.id.progress_bar).setVisibility(View.VISIBLE);
            ServerCommunicator.createAccount(this, ((QueuerApplication) getApplication()).getRequestQueue(),
                    userField.getText().toString(), passField.getText().toString(),
                    new Response.Listener<JSONObject>() {
                        boolean errored;
                        String errorText;

                        @Override
                        public void onResponse(JSONObject jsob) {
                            try {
                                Log.d("Received from /sessions", jsob.toString());
                                if (jsob.has("api_key") && jsob.getString("api_key") != null) {
                                    String apiKey = jsob.getString("api_key");
                                    ActivityUtils.saveApiKey(CreateAccountActivity.this, apiKey);
                                    //create user using Gson
                                    user = new Gson().fromJson(jsob.toString(), User.class);
                                    ActivityUtils.saveUserId(CreateAccountActivity.this, user.getId());
                                } else if (jsob.has("errors")) {
                                    //error
                                    errored = true;
                                    errorText = jsob.getString("errors");
                                } else {
                                    //error
                                    errored = true;
                                    errorText = "Unknown Error";
                                }
                            } catch (Exception e) {
                                e.printStackTrace();
                                //error
                                errored = true;
                                errorText = e.getLocalizedMessage();
                            }
                            new Handler().post(new Runnable() {
                                @Override
                                public void run() {
                                    findViewById(R.id.progress_bar).setVisibility(View.GONE);
                                    //go to the next activity
                                    if (!errored && user != null) {
                                        CreateAccountActivity.this
                                                .startActivity(new Intent(CreateAccountActivity.this, FeedActivity.class));
                                        CreateAccountActivity.this.finish();
                                    } else {
                                        ((TextView) findViewById(R.id.update)).setText(errorText);
                                    }
                                }
                            });
                        }
                    }, new Response.ErrorListener() {
                        @Override
                        public void onErrorResponse(final VolleyError volleyError) {
                            new Handler().post(new Runnable() {
                                @Override
                                public void run() {
                                    findViewById(R.id.progress_bar).setVisibility(View.GONE);
                                    ((TextView) findViewById(R.id.update)).setText(volleyError.getLocalizedMessage());
                                }
                            });
                        }
                    });
        }
    }
}

File: queuer/src/main/java/com/rndapp/task_feed/activities/CreateAccountActivity.kt
package com.rndapp.task_feed.activities

import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.EditText
import android.widget.TextView
import com.android.volley.Response
import com.google.gson.Gson
import com.rndapp.task_feed.R
import com.rndapp.task_feed.managers.SessionManager
import com.rndapp.task_feed.models.User
import org.json.JSONObject

/**
 * Created with IntelliJ IDEA.
 * User: ell
 * Date: 7/13/13
 * Time: 2:28 PM
 */
class CreateAccountActivity : AppCompatActivity(), View.OnClickListener {
    private var user: User? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_account)

        findViewById(R.id.create_account_button).setOnClickListener(this)
    }

    override fun onClick(view: View) {
        val userField = findViewById(R.id.username_field) as EditText
        val passField = findViewById(R.id.password_field) as EditText
        if (userField.text.toString() != "" && passField.text.toString() != "") {
            findViewById(R.id.progress_bar).visibility = View.VISIBLE
            SessionManager.createAccount(userField.text.toString(), passField.text.toString(),
                    object : Response.Listener<JSONObject> {
                        internal var errored: Boolean = false
                        internal var errorText: String = ""

                        override fun onResponse(jsob: JSONObject) {
                            try {
                                if (jsob.has("api_key") && jsob.getString("api_key") != null) {
                                    val apiKey = jsob.getString("api_key")
                                    SessionManager.saveApiKey(this@CreateAccountActivity, apiKey)
                                    //create user using Gson
                                    user = Gson().fromJson<User>(jsob.toString(), User::class.java!!)
                                    SessionManager.saveUserId(this@CreateAccountActivity, user!!.id)
                                } else if (jsob.has("errors")) {
                                    //error
                                    errored = true
                                    errorText = jsob.getString("errors")
                                } else {
                                    //error
                                    errored = true
                                    errorText = "Unknown Error"
                                }
                            } catch (e: Exception) {
                                e.printStackTrace()
                                //error
                                errored = true
                                errorText = e.localizedMessage
                            }

                            Handler().post {
                                findViewById(R.id.progress_bar).visibility = View.GONE
                                //go to the next activity
                                if (!errored && user != null) {
                                    this@CreateAccountActivity
                                            .startActivity(Intent(this@CreateAccountActivity, FeedActivity::class.java))
                                    this@CreateAccountActivity.finish()
                                } else {
                                    (findViewById(R.id.update) as TextView).text = errorText
                                }
                            }
                        }
                    }, Response.ErrorListener { volleyError ->
                Handler().post {
                    findViewById(R.id.progress_bar).visibility = View.GONE
                    (findViewById(R.id.update) as TextView).text = volleyError.localizedMessage
                }
            })
        }
    }
}
